plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java

```diff
@@ -0,0 +1,171 @@
+      Activator.logAndDisplayError(null, TITLE, e.getMessage());
+      return null;
+    }
+
+    if (!hasLoggedInUsers) {
+      Activator.logAndDisplayError(null,
+                                   TITLE,
+                                   "Please sign in to gcloud before deploying project "
+                                          + project.getName());
+      return null;
+    }
+
+    final IPath warLocation = getWarLocationOrPrompt(project);
+    if (warLocation == null) {
+      Activator.logAndDisplayError(null,
+                                   TITLE,
+                                   "Must select the WAR directory to deploy " + project.getName());
+      return null;
+    }
+
+    final IPath sdkLocation = runtime.getLocation();
+    if (sdkLocation == null) {
+      Activator.logAndDisplayError(null, TITLE, "Set the location of " + runtime.getId());
+      return null;
+    }
+
+    Job job = new Job("Running Cloud SDK Deploy Action") {
+      @Override
+      protected IStatus run(IProgressMonitor monitor) {
+        try {
+          List<String> commands = new ArrayList<String>();
+          commands.add(sdkLocation + GCloudCommandDelegate.GCLOUD_DIR);
+
+          commands.add("preview");
+          commands.add("app");
+          commands.add("deploy");
+          commands.add(warLocation.toString());
+
+          MessageConsole messageConsole = MessageConsoleUtilities.getMessageConsole(project.getName()
+                                                                                    + " - "
+                                                                                    + TITLE,
+                                                                                    null);
+          messageConsole.activate();
+
+          IPath projectLocation = project.getLocation();
+
+          int exitCode = ProcessUtilities.launchProcessAndWaitFor(commands,
+                                                                  projectLocation.toFile(),
+                                                                  messageConsole.newMessageStream(),
+                                                                  null);
+
+          if (exitCode != 0) {
+            Activator.logAndDisplayError(null,
+                                         TITLE,
+                                         "Cloud SDK deploy action terminated with exit code "
+                                                + exitCode);
+          }
+        } catch (IOException | InterruptedException e) {
+          Activator.logError(e);
+        }
+        return Status.OK_STATUS;
+      }
+    };
+    job.schedule();
+    return null;
+  }
+
+  private IPath getWarLocationOrPrompt(final IProject project) {
+    IFolder warDir = project.getFolder(WAR_SRC_DIR_DEFAULT);
+    if (warDir.exists()) {
+      return warDir.getLocation();
+    }
+
+    final IPath[] fileSystemPath = new IPath[1];
+    PlatformUI.getWorkbench().getDisplay().syncExec(new Runnable() {
```

Comment:
The activeShell should be passed in here, it's obtained from `HandlerUtil.getActiveShellChecked(event);`